io/io/src/TBufferMerger.cxx

```diff
@@ -90,37 +94,28 @@ void TBufferMerger::SetAutoSave(size_t size)
 
 void TBufferMerger::Merge()
 {
-   fBuffered = 0;
-   fMerger.PartialMerge();
-   fMerger.Reset();
-
-   if (fCallback)
-      fCallback();
-}
-
-void TBufferMerger::WriteOutputFile()
-{
-   std::unique_ptr<TBufferFile> buffer;
-
-   while (true) {
-      std::unique_lock<std::mutex> lock(fQueueMutex);
-      fDataAvailable.wait(lock, [this]() { return !this->fQueue.empty(); });
-
-      buffer.reset(fQueue.front());
-      fQueue.pop();
-      lock.unlock();
-
-      if (!buffer)
-         break;
+   std::queue<TBufferFile *> queue;
+   {
+      std::lock_guard<std::mutex> q(fQueueMutex);
+      std::swap(queue, fQueue);
+      fBuffered = 0;
+   }
+   {
+      std::lock_guard<std::mutex> m(fMergeMutex);
 
-      fBuffered += buffer->BufferSize();
-      fMerger.AddAdoptFile(new TMemFile(fMerger.GetOutputFileName(), buffer->Buffer(), buffer->BufferSize(), "read"));
+      while (!queue.empty()) {
+         std::unique_ptr<TBufferFile> buffer{queue.front()};
+         fMerger.AddAdoptFile(
```

Comment:
By using a queue and/or vector of TMemFile*, the 'new TMemFile' can be executed outside the fMergeMutex.
This is necessary as the TMemFile constructor can take the main ROOT write lock.

Reply:
The `new TMemFile` is inside a block with a lock_guard on the merge mutex (see line 104). How could it be run outside of the lock?

Reply:
```
std::deque<TMemFile*> memfiles;
while (!queue.empty()) {
   std::unique_ptr<TBufferFile> buffer{queue.front()};
   memfiles.pusb_back(new TMemFile(fMerger.GetOutputFileName(), buffer->Buffer(), buffer->BufferSize(), "READ"));
}
{
      std::lock_guard<std::mutex> m(fMergeMutex);
      while (!memfiles.empty()) {
         TMemFile *file{memfiles.front()};
         fMerger.AddAdoptFile(file);
         memfiles.pop();
      }
      fMerger.PartialMerge();
      fMerger.Reset();
}
```

Reply:
Not sure where you got the above code from, but it's not from the diff of my commits. Just look at the final diff.

Reply:
I wrote this code to illustrate what I mean with "How could it be run outside of the lock?".  The new TMemFile **must** be run outside of lock.  If it is run inside the local lock it will lead to dead-lock (because the TMemFile takes the ROOT global lock).

Reply:
I understand what you mean now. I thought you meant that it was outside the lock before. I have not seen deadlocks locally, but I agree that it looks like it could be a problem (and the cause of the test failure on mac). I will change the code to work as you are suggesting.